### ISSUE TITLE
Document overlay resources in the registry

### ISSUE THREAD

**Issue by t0yv0** [pulumi/pulumi#13231]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Currently resources implemented as provider overlays and available for a subset of languages do not have documentation entries in Pulumi Registry. For example, pulumi-gcp defines a gcp.cloudfunctions.HttpCallbackFunction Component resource. But it is not visible in https://www.pulumi.com/registry/packages/gcp/api-docs/cloudfunctions/ 

Note also that asking pulumi.ai questions about resources such as gcp.cloudfunctions.HttpCallbackFunction yields no examples at the moment; though this may have a different root-cause.

The suggested fix is to provide a mechanism to have the overlay resources documented in the registry for the languages that do support them (with overlay resources typically not all programming languages support the resource). Perhaps Pulumi Package Schema could have entries for these resources and carry their documentation.

**Comment by justinvp** [comment]
Yeah, I think we're going to want to add schema support the various types of overlay functionality we have, and then update docs-gen to properly emit documentation for it, for the languages that support it.

**Comment by t0yv0** [comment]
I had a quick conversation with @mjeffryes and wanted to add some color here, we need to be careful with scope as there is a fair amount of work here potentially.

K8S attempted to actually schematize overlays:
https://github.com/pulumi/pulumi-kubernetes/issues/1971

But then there are resources that resist it such as this one that are blocked from being schematized, and likely will be long-lived, and would definitely benefit from being officially documented:

https://github.com/pulumi/pulumi-kubernetes/pull/3020

I've just took an inventory of AWS Node overlays: 
https://github.com/pulumi/pulumi-aws/issues/4007

For AWS specifically, as discussed with @flostadler we would like to run an experiment to see if we can perchance schematize the serverless/lambda AWS ComponentResource overlays to make them MLCs and build upon them. 

There are more AWS overlays providing interfaces - I would like to understand better how it is used. Some of these patterns might be tricky to express generically in the schema.

**Comment by flostadler** [comment]
The possible avenues I see are documenting the overlays themselves and replacing them with native provider resources (e.g. MLCs) where applicable.

### tl;dr
We'll not be able to move all overlays to native resources, so we need a way to create documentation for them and customize the language chooser in the docs to account for unsupported languages.
A potential solution for that is leveraging the existing `IsOverlay` flag and adding the ability to customize the supported languages in the schema. This seems like a low hanging fruit that will allow us to document resources, types and free-standing functions.

Supporting resource methods (e.g. AWS `Bucket::onObjectRemoved`) in codegen will enable us to replace more overlays (and this way get them documented) but requires deeper changes to schema, codegen and docs.

For now I'd recommend to tackle the change to customize the language chooser for overlays because it's well contained and gives us immediate value. Then we can prioritize the support for resource methods later.

What do you think @mjeffryes @t0yv0 ?

### Document Overlays
To achieve this we could schematize overlays and generate the docs. There's a way to exclude types from codegen by marking them as an overlay (`IsOverlay` param on the type specs). The issue with that is that this would generate docs for all languages, while not all overlays are supported in every language (e.g. https://github.com/pulumi/pulumi-kubernetes/issues/2181). This creates confusion for users. We should add a mechanism for configuring what languages get displayed in the docs. That should be doable by hooking into the `LangChooser` here: https://github.com/pulumi/pulumi/blob/39c5647849857bd862814dce76136406f3967ecb/pkg/codegen/docs/gen.go#L363

The challenge with this approach is that not all overlays will be schematizable. Resource methods like `Bucket::onObjectRemoved` in AWS are not supported by the schema and codegen right now (https://github.com/pulumi/pulumi/issues/16257). This might sometimes involve having `any`-types because there's no way to schematize types like the callback input of the AWS classic `lambda.CallbackFunction`: `(event: E, context: Context, callback: (error?: Error | string | null, result?: R) => void) => Promise<R> | void`

### Replace overlays with native resources
Turning the overlays into native provider resources is not possible for all of them right now.
- The `lambda.CallbackFunction` overlay in AWS classic for example is specific to Node and we don't think there's meaningful alternatives in other languages (e.g. see https://github.com/pulumi/pulumi/issues/3820). For such resources we'd have to resort to schematizing and documenting the overlays.
- Some overlays are implemented as resource methods, but this is not supported by codegen right now. The **serverless connector mixins** (e.g. `Bucket::onObjectRemoved`) in AWS are implemented like that for example. We could invest in https://github.com/pulumi/pulumi/issues/16257 or hack together an injecting mechanism in the bridge, but both options would require changes to the docs pages to support those resource methods.

**Comment by flostadler** [comment]
While working on a prototype I found that we do support codegen/docsgen for resource methods (see https://github.com/pulumi/pulumi/issues/16257#issuecomment-2191701345). I tried whether I could docsgen docs for the existing resource methods. Generally it worked, but there were two problems:
- docs for the resource method get created for all languages the resource supports while the method is only present in a subset
- resource method docs do not support `MultiArgumentInputs` yet ([see](https://github.com/pulumi/pulumi/blob/39c5647849857bd862814dce76136406f3967ecb/pkg/codegen/docs/gen_method.go#L121)). It always generates a single object argument which doesn't allow matching the existing resource method overlays (e.g. [`Bucket.onObjectCreated`](https://github.com/pulumi/pulumi-aws/blob/8fab595268780f18c59c443cd993b6b9626b41d7/sdk/nodejs/s3/s3Mixins.ts#L223)).

**Comment by flostadler** [comment]
This has been shipped as part of [release v3.123.0](https://github.com/pulumi/pulumi/releases/tag/v3.123.0).

Enabling us to document resource method overlays will be tackled here: https://github.com/pulumi/pulumi/issues/16829
